Plugin.pm

```diff
@@ -467,18 +467,22 @@ sub QobuzArtist {
 		if ($artist->{albums}) {
 			my $albums = [];
 
-			my $sortByDate = (preferences('server')->get('jivealbumsort') || '') eq 'artflow';
+			# sort by release date if requested
+			my $sortByDate = $prefs->get('sortArtistAlbums');
 
 			$artist->{albums}->{items} = [ sort {
-
-				# push singles and EPs down the list
-				if ( ($a->{tracks_count} >= 4 && $b->{tracks_count} < 4) || ($a->{tracks_count} < 4 && $b->{tracks_count} >=4) ) {
-					return $b->{tracks_count} <=> $a->{tracks_count};
+				if ($sortByDate) {
+					return $sortByDate == 1 ? $b->{released_at}*1 <=> $a->{released_at}*1 : $a->{released_at}*1 <=> $b->{released_at}*1;
 				}
+				else {
+					# push singles and EPs down the list
+					if ( ($a->{tracks_count} >= 4 && $b->{tracks_count} < 4) || ($a->{tracks_count} < 4 && $b->{tracks_count} >=4) ) {
+						return $b->{tracks_count} <=> $a->{tracks_count};
+					}
 
-				return $a->{released_at}*1 <=> $b->{released_at}*1 if $sortByDate;
-				return lc($a->{title}) cmp lc($b->{title});
-
+					return lc($a->{title}) cmp lc($b->{title});
+				}
+				
 			} @{$artist->{albums}->{items} || []} ];
```

Comment:
That's certainly doing the job. 

If we were picky about performance, or dealing with large lists, you could optimize it: in its current state the condition for `$sortByDate` would be evaluated for each and every item. That's probably not much of a problem, as hardly any artist would have more than a few albums (besides Frank Zappa, who's got something between 100-200 on Spotify, IIRC...). But you could evaluate it outside the loop, then create an anonymous sub to execute the sorting comparison. Something like:


```perl
my $sorter;

if ($sortByDate) {
   if ($sortByDate == 1) {
      $sorter = sub { $b->{released_at}*1 <=> $a->{released_at} };
   } else {
      $sorter = sub { $a->{released_at}*1 <=> $b->{released_at} };
   }
}
else {
   $sorter = sub {
      if ( ($a->{tracks_count} >= 4 ...) {
         return $b->{tracks_count} <=> $a->{tracks_count};
      }

      return lc($a->{title})...;   
   };
}

$artist->{albums}->{items} = [ sort $sorter @{$artist->{albums}->{items} || []} ];
```

I guess you get the idea. But as I said: that's just an idea, not really needed here (nor fully implemented or tested).

Reply:
Good point. I'm usually big on code efficiency but this is my first experience with sorting in Perl and I admit there was some cutting and pasting going on here.  ;-)